container-search/src/main/java/com/yahoo/search/ranking/SimpleEvaluator.java

```diff
@@ -0,0 +1,48 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.search.ranking;
+
+import ai.vespa.models.evaluation.FunctionEvaluator;
+import com.yahoo.search.result.FeatureData;
+import com.yahoo.search.result.Hit;
+import com.yahoo.tensor.Tensor;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class SimpleEvaluator implements Evaluator {
+
+    private final FunctionEvaluator evaluator;
+    private final Set<String> neededInputs;
+
+    
+    public SimpleEvaluator(FunctionEvaluator prototype) {
+        this.evaluator = prototype;
+        this.neededInputs = new HashSet<String>(prototype.function().arguments());
```

Comment:
The convention in Java is to perform defence copying in constructor for immutable types. That will remove the need for `List.copyOf` for each `needInputs()` invocation.